perf: Use a single node to represent multiple shadow variables in graphs

Christopher-Chianelli opened the pull request:
When there is only a single declarative entity and all non-declarative parents go in the same direction (ex: previous), then some variables can share the same node in the graph. In particular, there is a static topological order when only the variables on a single entity is considered, and as long as the next declarative shadow variable does not have any new non-declarative parents (i.e. group), it can be grouped into the same node as the previous one. All nodes share the same node if no groups are used.

Also fixed a bug in the topological sort done to determine variable trigger order when groups are used (previously it added group variables, which caused it to create a cycle and make all topological orders valid).

Christopher-Chianelli committed:
perf: Use a single node to represent multiple shadow variables in graphs

When there is only a single declarative entity and all
non-declarative parents go in the same direction (ex: previous),
then some variables can share the same node in the graph. In
particular, there is a static topological order when only the
variables on a single entity is considered, and as long as
the next declarative shadow variable does not have any new
non-declarative parents (i.e. group), it can be grouped into
the same node as the previous one.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/AffectedEntitiesUpdater.java:
```diff
@@ -127,21 +127,26 @@ private boolean updateShadowVariable(EntityVariablePair<Solution_> entityVariabl
-                return true;
+        var anyChanged = false;
+        for (var shadowVariableReference : shadowVariableReferences) {
+            var oldValue = shadowVariableReference.memberAccessor().executeGetter(entity);
+            if (isLooped) {
+                if (oldValue != null) {
+                    affectedEntities.add(entityVariable);
+                    changeShadowVariableAndNotify(shadowVariableReference, entity, null);
+                }
+                anyChanged = true;
+            } else {
+                var newValue = shadowVariableReference.calculator().apply(entity);
+                if (!Objects.equals(oldValue, newValue)) {
+                    affectedEntities.add(entityVariable);
+                    changeShadowVariableAndNotify(shadowVariableReference, entity, newValue);
+                    anyChanged = true;
+                }
             }
```
Please extract this to a new function. Will make profiling easier in the future.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultShadowVariableSessionFactory.java:
```diff
@@ -143,7 +147,7 @@ private static <Solution_> VariableReferenceGraph buildArbitraryGraph(
             IntFunction<TopologicalOrderGraph> graphCreator) {
         var declarativeShadowVariableDescriptors = solutionDescriptor.getDeclarativeShadowVariableDescriptors();
-        var variableIdToUpdater = new HashMap<VariableMetaModel<?, ?, ?>, VariableUpdaterInfo<Solution_>>();
+        var variableIdToUpdater = new HashMap<VariableMetaModel<?, ?, ?>, VariableUpdaterInfo<Solution_>[]>();
```
IMO this is complex enough now to warrant its own type.
Even if it just wraps the map, the names of the methods and fields of the new type will make it more obvious what's happening here.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/GraphStructure.java:
```diff
@@ -99,12 +128,33 @@ public static <Solution_> GraphStructureAndDirection determineGraphStructure(
+        }
+    }
+
+    private static <Solution_> boolean doesEntitiesUseDeclarativeShadowVariables(Object[] entities,
```
`doEntities...`?
Also, maybe switch the order of the arguments, and make entities a vararg?

triceo reviewed (approved):
LGTM, small comments.

triceo commented:
Oh and also, a green Sonar rating, please.

Christopher-Chianelli committed:
perf: review comments, all nodes share the same node if no groups are used

- Also fixed a bug in the topological sort done to determine variable
  trigger order when groups are used (previously it added group variables,
  which caused it to create a cycle and make all topological orders
  valid).

triceo committed:
Sonar and small cleanup
